WordPress/src/main/java/org/wordpress/android/ui/reader/repository/ReaderPostRepository.kt

```diff
@@ -1,335 +1,149 @@
-import org.wordpress.android.datasets.ReaderTagTable
 import org.wordpress.android.models.ReaderPostList
 import org.wordpress.android.models.ReaderTag
 import org.wordpress.android.modules.BG_THREAD
-import org.wordpress.android.modules.IO_THREAD
-import org.wordpress.android.ui.reader.ReaderConstants
-import org.wordpress.android.ui.reader.ReaderEvents.SearchPostsEnded
-import org.wordpress.android.ui.reader.ReaderEvents.SearchPostsStarted
 import org.wordpress.android.ui.reader.ReaderEvents.UpdatePostsEnded
-import org.wordpress.android.ui.reader.ReaderEvents.UpdatePostsStarted
-import org.wordpress.android.ui.reader.actions.ReaderActions.UpdateResult.CHANGED
-import org.wordpress.android.ui.reader.actions.ReaderActions.UpdateResult.FAILED
-import org.wordpress.android.ui.reader.actions.ReaderActions.UpdateResult.HAS_NEW
-import org.wordpress.android.ui.reader.actions.ReaderActions.UpdateResult.UNCHANGED
-import org.wordpress.android.ui.reader.services.post.ReaderPostServiceStarter
-import org.wordpress.android.ui.reader.services.post.ReaderPostServiceStarter.UpdateAction
-import org.wordpress.android.ui.reader.services.post.ReaderPostServiceStarter.UpdateAction.REQUEST_NEWER
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryCommunication.Error.RemoteRequestFailure
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryUseCaseType.FETCH_NUM_POSTS_BY_TAG
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryUseCaseType.FETCH_POSTS_BY_TAG
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryUseCaseType.FETCH_POSTS_BY_TAG_WITH_COUNT
+import org.wordpress.android.ui.reader.repository.ReaderRepositoryUseCaseType.SHOULD_AUTO_UDPATE_TAG
 import org.wordpress.android.util.EventBusWrapper
 import org.wordpress.android.util.NetworkUtilsWrapper
 import org.wordpress.android.viewmodel.ContextProvider
 import org.wordpress.android.viewmodel.Event
 import org.wordpress.android.viewmodel.ReactiveMutableLiveData
 import javax.inject.Inject
 import javax.inject.Named
-import kotlin.coroutines.CoroutineContext
 
-class ReaderPostRepository
-constructor(
-    private val contextProvider: ContextProvider,
+class ReaderPostRepository(
+    private val bgDispatcher: CoroutineDispatcher,
     private val eventBusWrapper: EventBusWrapper,
     private val networkUtilsWrapper: NetworkUtilsWrapper,
-    private val bgDispatcher: CoroutineDispatcher,
-    private val ioDispatcher: CoroutineDispatcher,
-    private val readerTag: ReaderTag
-) : CoroutineScope {
+    private val contextProvider: ContextProvider,
+    private val readerTag: ReaderTag,
+    private val fetchPostsForTagUseCase: FetchPostsForTagUseCase,
+    private val fetchNumPostsForTagUseCase: FetchNumPostsForTagUseCase,
+    private val shouldAutoUpdateTagUseCase: ShouldAutoUpdateTagUseCase,
+    private val fetchPostsForTagWithCountUseCase: FetchPostsForTagWithCountUseCase
+) : BaseReaderRepository(eventBusWrapper,
```

Comment:
Have you considered using composition instead of inheritance here? It seems we could get away with composition which is imo always better. Wdyt?

Reply:
100% agree.

Reply:
@malinajirka - I've started the extraction away from the BaseReaderRepository for `ReaderDiscoverRepository` in commit 3b5e54a. Take a look at `ReaderUpdatePostsEndedHandler` (and yes, I just realized it landed in the usecase folder - I'll move that) and let me know your thoughts. Is this overkill or does it nicely separate? Thanks

Reply:
I personally like it :), gj!